### ISSUE TITLE
What to do about old versions of the docs on readthedocs?

### ISSUE THREAD

**Issue by nicoddemus** [pytest-dev/pytest#8098]
Yeah recently we had a few cases of misunderstandings because of old links. 

Usually we only really keep two docs up-to-date, which is `stable` and `latest`.

What should we do with old documentation pages? 

Btw the docs mentioned in the original issue are quite old, from the `reorganize-docs` branch, back in 2016. I don't think we will really pick it up again at this point, perhaps we can just delete them to avoid confusion? @pfctdayelise @hackebrot?

_Originally posted by @nicoddemus in https://github.com/pytest-dev/pytest/issues/8094#issuecomment-738465588_

**Comment by pfctdayelise** [comment]
Good idea to avoid confusion, thanks @nicoddemus !

**Comment by RonnyPfannschmidt** [comment]
btw,  do  we still have the notes from back then, i believe we should try to make what we set out back  in the sprint possible in a more iterative small pieced fashion

**Comment by pfctdayelise** [comment]
Hmm...the broad idea I believe was to rearrange the docs to distinguish between
- Tutorial
- How-to guide
- Discussions
- Reference

This is described by Daniel Procida in a talk https://pyvideo.org/pycon-au-2017/what-nobody-tells-you-about-documentation.html (although apparently that was 2017, so maybe my brain just made this up....)

Our notes about it are here https://github.com/pytest-dev/pytest/wiki/Docs-refactor , where we didn't mention that, so maybe we were just planning to organise a bit and categorise according to level of user. I do think the docs are still a bit overwhelming and there is not that much navigation help. The reference material has improved a bit though :) not sure if it is complete yet!

**Comment by aklajnert** [comment]
Since the users are redirected to the old docs from Google, maybe instead of removing them, it would be a good idea to redirect to the new one or display a huge banner, that these docs are for the old version with a link to the new one?

**Comment by nicoddemus** [comment]
> Since the users are redirected to the old docs from Google, maybe instead of removing them, it would be a good idea to redirect to the new one or display a huge banner, that these docs are for the old version with a link to the new one?

Indeed that's ideal, however that setting is already configured:

![image](https://user-images.githubusercontent.com/1085180/101161842-39adcc00-3610-11eb-8f6f-a76f7f207ed3.png)

Perhaps the banner is not showing up because of our theme?

**Comment by Zac-HD** [comment]
For [Hypothesis](https://hypothesis.readthedocs.io/en/latest/), we only host "latest", specificially so that stale docs can't show up in search results etc., and I'd support "only stable, latest, and 4.x" for pytest for the same reasons.  

*(we're also hosting the last py2 docs for now, but free support only applies to the latest release and nobody pays, so :man_shrugging:)*

**Comment by nicoddemus** [comment]
>  I'd support "only stable, latest, and 4.x" for pytest for the same reasons.

👍 from me. Does anybody see a good reason to host other versions?

**Comment by asottile** [comment]
> 
> 
> > I'd support "only stable, latest, and 4.x" for pytest for the same reasons.
> 
> 👍 from me. Does anybody see a good reason to host other versions?

sg2m!

**Comment by Zac-HD** [comment]
Sounds like all we need is for someone with edit access (e.g. @nicoddemus, @pfctdayelise, @RonnyPfannschmidt, @asottile) to go to https://readthedocs.org/projects/pytest/versions/ while logged in, then `edit -> uncheck active -> save` for all the older versions.  I'd be happy to but don't have write access.

**Comment by asottile** [comment]
whew that was a lot of clicking -- but I think it's done

**Comment by Zac-HD** [comment]
![image](https://user-images.githubusercontent.com/12229877/103330223-a2278600-4ab4-11eb-9546-002cae0f02f6.png)

🎉

**Comment by PhilippSelenium** [comment]
Imagine the following scenario:
I have a working sphinx and CI build where pytest is pinned to i.e. 6.2.0
Now pytest does a major version upgrade (i.e. 7.0.0) (and some interfaces were changed/removed/...)
Despite having pytest pinned my sphinx build will break :(
And I have no option but to update to the newest version ...

**Comment by Zac-HD** [comment]
I'm sorry that this is a potentially breaking change for you 😢 

However we do think that the reduced risk of confusion for less experienced users outweighs the *potential* breaking change for your docs.  You can also work around this by saving a local copy, building Pytest docs from a git submodule, etc., whereas there's no equivalent workaround for the problem of outdated documentation appearing in search results.

**Comment by PhilippSelenium** [comment]
@Zac-HD I think I will have to  ... the needs of the many ... ;)
Thanks for the hint about the workaround we are already doing that.

**Comment by RonnyPfannschmidt** [comment]
Read the docs could perhaps use a add on to help referring to current docs and making searches ignore outdated docs

**Comment by The-Compiler** [comment]
It looks like the `reorganize-docs` branch is still around on RTD and apparently turns up in Google search result (I just had a coworker ask why the snippet from the docs wouldn't work). Is there something we can do about this?

**Comment by RonnyPfannschmidt** [comment]
Let's take a look at dropping, once I get back to the computer

**Comment by Zac-HD** [comment]
Just checking in on this - [it's gone from readthedocs](https://readthedocs.org/projects/pytest/), but [still available on pytest.org](https://docs.pytest.org/en/reorganize-docs/) - and causing trouble in e.g. #8655.

Possibly stale files on a server somewhere, since the readthedocs version selector is missing too?

**Comment by The-Compiler** [comment]
Well, docs.pytest.org points to readthedocs.io:

```
$ dig docs.pytest.org A     
[...]

;; ANSWER SECTION:
docs.pytest.org.	262	IN	CNAME	readthedocs.io.
readthedocs.io.		262	IN	A	104.17.32.82
readthedocs.io.		262	IN	A	104.17.33.82
```

so either we're missing something from those settings, or it's a bug on their side...

**Comment by The-Compiler** [comment]
I've tried adding a prefix redirect /en/reorganize-docs/ -> /en/6.2.x/ in RTD as a workaround, but looks like that didn't change anything either (so I now removed it again)...

**Comment by The-Compiler** [comment]
FYI, I opened https://github.com/readthedocs/readthedocs.org/issues/8264 for this now (cc @kini)

**Comment by RonnyPfannschmidt** [comment]
@The-Compiler if all fails, how about recreating the branch once from main and dropping it once regendoc can actually  see it  and regenerated it again (only if they cant fix the issue timely)

**Comment by The-Compiler** [comment]
I assume you mean rtd, not regendoc? Either way, seems like a very temporary solution, but perhaps worth a try. On the other hand, we've been living with this for years, so it's not like another week or two makes a difference, as long as it gets fixed.

**Comment by The-Compiler** [comment]
The good news: I got things sorted out, the old docs are gone, and going to e.g. [reorganize-docs/index.html](https://docs.pytest.org/en/reorganize-docs/index.html) will redirect to [6.2.x/index.html](https://docs.pytest.org/en/6.2.x/index.html).

The bad news: It falls apart for pages which have a redirect due to https://github.com/pytest-dev/pytest/issues/8471#issuecomment-872421119 - for example, [reorganize-docs/mark.html](https://docs.pytest.org/en/reorganize-docs/mark.html) will incorrectly redirect to [6.2.x/how-to/mark.html](https://docs.pytest.org/en/6.2.x/how-to/mark.html) which is still a 404.

However, this should fix itself once we got 7.0 released, as at that point we'll likely switch back to `/stable/` / `/latest/` for the docs, and the docs will exist at the new path (with `how-to`).

Still I've reported this to RTD here: https://github.com/readthedocs/readthedocs.org/issues/8309
